state-chain/pallets/cf-elections/src/electoral_systems/block_witnesser/block_processor.rs

```diff
@@ -203,57 +181,62 @@ impl<T: BWProcessorTypes> BlockProcessor<T> {
 		last_events
 	}
 
-	/// Applies the processing rules for a given block at a specific age to generate events.
+	/// Applies the processing rules for a given block and a given range of ages to generate events.
 	///
 	/// This function performs two primary steps:
 	///
-	/// 1. **Event Generation:** It calls the `rules` hook with a tuple `(block, age, data.clone())`
-	///    to generate events.
+	/// 1. **Event Generation:** It calls the `rules` hook with a tuple `(block_number, age, data,
+	///    safety_margin)` to generate events.
 	/// 2. **Deduplication Filtering:** It then filters out events that are already present in
-	///    `reorg_events`
+	///    `processed_events`. If an event is already present in `processed_events`, it is
+	///    considered a duplicate. The existing entry is updated to reflect the highest expiry
+	///    ChainBlockNumber between the existing and the new (duplicate) event.
 	///
 	/// # Parameters
 	///
-	/// - `block`: The block number for which to process rules.
+	/// - `block_number`: The block number for which to process rules.
 	/// - `age`: The age of the block (i.e., how many blocks have passed since this block).
 	/// - `data`: A reference to the block data.
+	/// - `safety_margin`: the safety margin for that block
 	///
 	/// # Returns
 	///
 	/// A vector of (block height, events (`T::Event`)) generated by applying the rules, excluding
-	/// any duplicates.
+	/// any event already processed.
 	fn process_rules_for_ages_and_block(
 		&mut self,
-		block: T::ChainBlockNumber,
+		block_number: T::ChainBlockNumber,
 		age: Range<u32>,
 		data: &T::BlockData,
 		safety_margin: u32,
 	) -> Vec<(T::ChainBlockNumber, T::Event)> {
 		let events: Vec<(T::ChainBlockNumber, T::Event)> =
-			self.rules.run((block, age, data.clone(), safety_margin));
+			self.rules.run((block_number, age, data.clone(), safety_margin));
+
 		events
 			.into_iter()
-			.filter(|(_, last_event)| {
-				!self
-					.reorg_events
-					.iter()
-					.flat_map(|(_, (events, _))| events)
-					.any(|event| event == last_event)
+			.filter(|(block_number, event)| {
+				let expiry = block_number.saturating_forward(safety_margin as usize);
+				match self.processed_events.get_mut(event) {
+					Some(stored_expiry) => {
+						if *stored_expiry < expiry {
+							*stored_expiry = expiry;
+						}
+						false
+					},
+					None => {
+						self.processed_events.insert(event.clone(), expiry);
+						true
+					},
+				}
```

Comment:
It's a bit surprising that this function mutates state, because it feels like one might to want to call it without the mutation effect. If that's not true then I'm fine with keeping it this way, though I would try to update the name of the function to make clear that it is mutating / has side effects.

Thoughts @kylezs ?